p2p/src/util/src/lib.rs

```diff
@@ -0,0 +1,191 @@
+        let inner = retval.inner;
+
+        (
+            quote! {
+                #ident
+            },
+            quote! {
+                #inner
+            },
+        )
+    })
+}
+
+#[proc_macro_derive(Handle, attributes(name, return_value))]
+pub fn event(input: TokenStream) -> TokenStream {
+    let data = parse_macro_input!(input as DeriveInput);
+    let orig = data.ident.clone();
+    let sender = syn::Ident::new(format!("{}Sender", data.ident).as_str(), Span::call_site());
+
+    let filtered = if let syn::Data::Enum(syn::DataEnum { variants, .. }) = data.data {
+        variants
+            .into_pairs()
+            .filter_map(|x| match x {
+                Punctuated(entry, _) => Some(entry),
+                _ => None,
+            })
+            .collect::<Vec<_>>()
+    } else {
+        panic!("`#[derive(Handle)]` can only be used with enums!");
+    };
+
+    let methods = filtered
+        .iter()
+        .map(|f| {
+            let name = &f.ident;
+            let fields = &f.fields;
+
+            let real_name = parse_real_name(&f.attrs).unwrap_or_else(|| {
+                let name = syn::Ident::new(&name.to_string().to_case(Case::Snake), Span::call_site());
+                quote! { #name }
+            });
+
+            let (func_args, chan_args): (Vec<_>, Vec<_>) = fields
+                .iter()
+                .map(|field| {
+                    let ident = &field.ident;
+                    let ty = &field.ty;
+
+                    (!field
+                        .attrs
+                        .iter()
+                        .any(|attr| attr.path.is_ident("return_value")))
+                    .then_some((
+                        quote! {
+                            , #ident: #ty
+                        },
+                        quote! {
+                            #ident
+                        },
+                    ))
+                })
+                .flatten()
+                .unzip();
+
+            if let Some((retval_name, retval_type)) = parse_return_value(fields) {
+                quote! {
+                    pub async fn #real_name(&mut self #(#func_args)*) -> core::result::Result<#retval_type, P2pError> {
+                        let (tx, rx) = oneshot::channel();
+                        self.tx
+                            .send(#orig::#name { #(#chan_args,)* #retval_name: tx })
+                            .map_err(|_| P2pError::ChannelClosed)?;
+                        rx.await.map_err(|_| P2pError::ChannelClosed)
+                    }
+                }
+            } else {
+                quote! {
+                    pub async fn #real_name(&mut self #(#func_args)*) -> core::result::Result<(), P2pError> {
+                        self.tx
+                            .send(#orig::#name { #(#chan_args,)* })
+                            .map_err(|_| P2pError::ChannelClosed)
+                    }
+                }
+            }
```

Comment:
I am somewhat uncomfortable that the presence or absence of return value determines whether the call blocks the current task. If someone decides to add or remove the return value, the behaviour here changes silently.

Not sure what a good approach may be here. Maybe prefix the method name with `send_` for fire-and-forget or with `call_` when waiting for the result is requested. Not sure if the `send_` version should only be generated for things with no return type or for everything. Ergonomics may suffer a bit but at least it's readily obvious which one is used by looking at the call. Any thoughts?

Reply:
Now that I removed the unnecessary `async`, the split is actually quite obvious. Fire-and-forget function calls are synchronous whereas functions with return value require you to `await` on them. Do you think this is clear enough distinction? I personally don't like the idea of prefixing them with something but if others also think it's better to have a prefix I don't mind adding one too much.

Reply:
> Now that I removed the unnecessary `async`, the split is actually quite obvious. Fire-and-forget function calls are synchronous whereas functions with return value require you to `await` on them. Do you think this is clear enough distinction? I personally don't like the idea of prefixing them with something but if others also think it's better to have a prefix I don't mind adding one too much.

Yea presence or absence of `.await` is a clear enough distinction I think.